### ISSUE TITLE
wrap-iife doesn't catch .call / .apply. JSCS does

### ISSUE THREAD

**Issue by TheSavior** [eslint/eslint#7212]
**Tell us about your environment**

* **ESLint Version:** 3.5.0
* **Node Version:** N/A
* **npm Version:** N/A

**What parser (default, Babel-ESLint, etc.) are you using?** Default

**Please show your full configuration:**
```js
/* global window */
/* eslint wrap-iife: ["error", "inside"] */
window.foo = (function() { return 1; })();
window.bar = function() { return 3; }.call(this, arg1);
window.baz = function() { return d; }.apply(this, args);
```

The following passes. window.foo is fine because it is wrapped in parens, but the others don't require it.

I'd expect that the functions that are called using `.call` and `.apply` should also be wrapped in parens. That is the behavior of the corresponding rule in JSCS and I'd like to keep that behavior while converting: http://jscs.info/rule/requireParenthesesAroundIIFE

## Proposal 
The current proposal is to add a new boolean option called `functionPrototypeMethods` that will lint `call` and `apply` using the same `inside` or `outside` value. For example:

```
/* eslint wrap-iife: [2, "outside", { functionPrototypeMethods: true }] */

(function(){ foo(); }()) // ok
(function(){ foo(); }.call(bar)) // ok

(function(){ foo(); }).call(bar) // error, expected parens to be outside but they were inside
```

**Comment by kaicataldo** [comment]
This sounds like a reasonable request to me, since it's something the corresponding JSCS rule checks. Seems like it should be treated as an enhancement of the rule and enabled with an option.

**Comment by TheSavior** [comment]
Any thoughts on what the option should be called?

**Comment by TheSavior** [comment]
I imagine this should also require `.bind` to be wrapped. So it seems like the two things this rule needs to handle are `FunctionExpression` and `MemberExpression` inside of a `CallExpression`. 

In order to maintain backwards compatibility, assuming that is important, is to have the following schema:


```
wrap-iife: ["error", {
  CallExpression: "inside",
  MemberExpression: "outside"
}]
```

This current schema:
```
wrap-iife: ["error", "outside"]
```
would be the same as:

```
wrap-iife: ["error", {
  CallExpression: "outside"
}]
```

The problem with this to me is that I'd imagine most people would want and expect to be linting for all of these cases and so we would be making the standard case require additional configuration.

I'd much prefer to make a non backwards compatible change resulting in the current schema:

```
wrap-iife: ["error", "outside"]
```
translating to

```
wrap-iife: ["error", {
  CallExpression: "outside",
  MemberExpression: "outside"
}]
```

Thoughts?

**Comment by not-an-aardvark** [comment]
I'm not sure I agree with requiring `bind` to be wrapped, since at that point it's not really an IIFE. But if other people want it or it's required for JSCS compatibility, then I could live with it.

Other possibilities for the option name:

* `{call: true, apply: true}` (both false by default, `true` indicates that parens are required for these cases
* Maybe something like `functionPrototypeInvocations` if we want to only have one option?

**Comment by TheSavior** [comment]
Im not tied to supporting bind either. Do we still need to support inside and outside on call and apply?

**Comment by not-an-aardvark** [comment]
I was assuming that people would want to use the same config for .call/.apply as regular IIFEs. For example:

```js
/* eslint wrap-iife: [2, "outside", {"call": true, "apply": true}] */

(function(){ foo(); }()) // ok
(function(){ foo(); }.call(bar)) // ok

(function(){ foo(). }).call(bar) // error, expected parens to be outside but they were inside
```

**Comment by TheSavior** [comment]
Yes, totally agree. I think people would want the same behavior for all of them. That is the ideal change. 

How do we feel about it breaking the backwards compatibility? Are we okay with that?

**Comment by TheSavior** [comment]
That sounds good to me. Are we able to break backwards compatibility and set those true by default? Or do we frown on that and make them false by default?

**Comment by not-an-aardvark** [comment]
I think we would make them false by default (at least for now), since making them true by default would be a semver-major change. We could consider changing the default next time a major release happens.

**Comment by TheSavior** [comment]
Great. Happy to implement this change once the ticket gets accepted.

**Comment by not-an-aardvark** [comment]
Sounds good, I'll champion the proposal.

I'm not sure `call` and `apply` are ideal option names (it might not be immediately clear what they do), but if we think of a better option name before the change gets merged, it shouldn't be too hard to update.

**Comment by TheSavior** [comment]
Great to hear.

If you aren't happy with `call` and `apply` (I'm cool with those), what about a single option called `MemberExpression`?

```
/* eslint wrap-iife: [2, "outside", { MemberExpression: true }] */
```

**Comment by kaicataldo** [comment]
I'm :+1: for this change, but not sure about the option name...I think it makes sense to have a single option over two - seems like if you want this rule to apply to `call` or `apply` you'd also want it to apply to the other.

`MemberExpression` isn't clear enough for my taste. I'll throw another one out there - `functionPrototypeMethods`?

**Comment by TheSavior** [comment]
I don't feel strongly about any of these. I mostly just want to get this behavior out. :P

Current (not shot down) proposals are:

`functionPrototypeMethods` and `functionPrototypeInvocations`. Anyone else have thoughts?

**Comment by kaicataldo** [comment]
Hope it doesn't actually feel like we're shooting your ideas down - that's not the intention!

**Comment by TheSavior** [comment]
Nope nope, it's fine. I legitimately don't care, just want to help push these forward. :) I'm cool with any option name

**Comment by not-an-aardvark** [comment]
I prefer `functionPrototypeMethods` over `functionPrototypeInvocations`.

**Comment by TheSavior** [comment]
Great, sounds like the three of us are good with `functionPrototypeMethods`. What's the next step?

**Comment by kaicataldo** [comment]
@TheSavior Really appreciate you staying on this :) Mind updating the original issue to include the proposed option name as well? Next step is for the the champion to gain consensus (3 supporters) from the team. I've put my thumb on it, so we just need two more to support this addition.

**Comment by TheSavior** [comment]
Updated the initial post with the current proposal

**Comment by not-an-aardvark** [comment]
@TheSavior This issue is accepted now, so feel free to start implementing this (assuming you're still interested in doing so).

**Comment by TheSavior** [comment]
Sure, will do. This should be pretty straight forward. (Famous last words)
